riot.py

```diff
@@ -0,0 +1,480 @@
+                    ("python-memcached", [""]),
+                    ("django", [">=2.0,<2.1", ">=2.1,<2.2"]),
+                ],
+            ),
+            Case(
+                env=[("TEST_DATADOG_DJANGO_MIGRATION", [None, "1"])],
+                pys=[3.6, 3.7, 3.8],
+                pkgs=[
+                    ("django-pylibmc", [">=0.6,<0.7"]),
+                    ("django-redis", [">=4.5,<4.6"]),
+                    ("pylibmc", [""]),
+                    ("python-memcached", [""]),
+                    ("django", [">=2.0,<2.1", ">=2.1,<2.2", ">=2.2,<2.3", ">=3.0,<3.1"]),
+                ],
+            ),
+        ],
+    ),
+]
+
+
+def rmchars(chars: t.List[str], s: str):
+    for c in chars:
+        s = s.replace(c, "")
+    return s
+
+
+def get_pep_dep(libname: str, version: str):
+    """Returns a valid PEP 508 dependency string.
+
+    ref: https://www.python.org/dev/peps/pep-0508/
+    """
+    return "%s%s" % (libname, version)
+
+
+def get_env_str(envs: t.List[t.Tuple]):
+    return " ".join("%s=%s" % (k, v) for k, v in envs)
+
+
+def get_base_venv_path(pyversion):
+    """Given a python version return the base virtual environment path relative
+    to the current directory.
+    """
+    return ".venv_py%s" % str(pyversion).replace(".", "")
+
+
+def run_cmd(*args, **kwargs):
+    # Provide our own defaults.
+    if "shell" in kwargs and "executable" not in kwargs:
+        kwargs["executable"] = SHELL
+    if "encoding" not in kwargs:
+        kwargs["encoding"] = ENCODING
+    if "stdout" not in kwargs:
+        kwargs["stdout"] = subprocess.PIPE
+
+    logger.debug("Running command %s", args[0])
+    r = subprocess.run(*args, **kwargs)
+    logger.debug(r.stdout)
+
+    if r.returncode != 0:
+        raise CmdFailure("Command %s failed with code %s." % (args[0], r.returncode), r)
+    return r
+
+
+def create_base_venv(pyversion, path=None, recreate=True):
+    """Attempts to create a virtual environment for `pyversion`.
+
+    :param pyversion: string or int representing the major.minor Python
+                      version. eg. 3.7, "3.8".
+    """
+    path = path or get_base_venv_path(pyversion)
+
+    if os.path.isdir(path) and not recreate:
+        logger.info("Skipping creating virtualenv %s as it already exists.", path)
+        return path
+
+    py_ex = "python%s" % pyversion
+    py_ex = shutil.which(py_ex)
+
+    if not py_ex:
+        logger.debug("%s interpreter not found", py_ex)
+        raise FileNotFoundError
+    else:
+        logger.info("Found Python interpreter '%s'.", py_ex)
+
+    logger.info("Creating virtualenv '%s' with Python '%s'.", path, py_ex)
+    r = run_cmd(["virtualenv", "--python=%s" % py_ex, path], stdout=subprocess.PIPE)
+    return path
+
+
+def get_venv_command(venv_path, cmd):
+    """Return the command string used to execute `cmd` in virtual env located
+    at `venv_path`.
+    """
+    return "source %s/bin/activate && %s" % (venv_path, cmd)
```

Comment:
you probably don't _have_ to call `source`. If you read through `activate` majority of it is setting up `deactivate`, and then setting a few environment variables:

```
export VIRTUAL_ENV=<path_to_venv>
export PATH=<venv>/bin:$PATH
unset PYTHONHOME
```

And that is in.

In fact, you can generally get by just calling from `<venv>/bin/<cmd>` directly without modifying any env variables and it'll work as expected.

Reply:
Calling `source .../bin/activate` is _probably_ the right thing to do